Expand and clarify JBP-7 (#1)
JonathanGiles/java-best-practices: docs/java-dev-best-practices/jbp-0007.md

jbp:
  id: JBP-7
permalink: /JBP-7
title: Use the latest Java Long-Term Support release
---

{% include links.md %}

> New Java releases use a six-month feature-release-train model rather than having a major release planned for every two to four years, with this rapid release cadence, Oracle and other participants in the Java ecosystem designated *Long-Term Support* (or LTS) releases, these LTS releases are made every three years, with the most recent LTS releases being Java 11 (September 2018) and Java 17 (September 2021).

Reviewer: ```suggestion
Java releases are made every six months in a 'feature-release-train' model, rather than having a major release planned for every two to four years as historically was the case. This rapid release cadence enables Oracle, and other participants in the Java ecosystem, to designate *Long-Term Support* (or LTS) releases every three years, with the most recent LTS releases being Java 11 (September 2018) and Java 17 (September 2021).
```

> The major difference between non-LTS releases and LTS releases is that when a new release is made available (every six months), the previous non-LTS is out-of-support, this means that it won't receive quarterly security, stability, or performance updates. Most organizations still prefer to target LTS versions for production use because is not always possible to test and update the JDK every six months.

Reviewer: ```suggestion
The major difference between non-LTS releases and LTS releases is that when a new release is made available (every six months), the previous non-LTS is out-of-support. When the release leaves support, it meanes that it won't receive quarterly security, stability, or performance updates any longer. Most organizations still prefer to target LTS versions for production use because it is not always possible to test and update their production JDK every six months.
```

> Non‑LTS releases are considered a cumulative set of implementation enhancements of the most recent LTS release. These are good releases to test with so that you can have higher degrees of confidence that moving to the next LTS will be straightforward, but you should avoid using non-LTS releases in production unless you can test and update the JDK with every new release.

Reviewer: ```suggestion
Non‑LTS releases are considered a cumulative set of enhancements of the most recent LTS release. These are good releases to test with so that you can have higher degrees of confidence that moving to the next LTS will be straightforward, but you should avoid using non-LTS releases in production unless you can test and update the JDK with every new release.
```

## Keeping the JDK up to date

> The main recomendation is that the JDK is updated with each Patch Update, when using an LTS release you should always test and update to the latest patch which contain security vulnerability fixes and other improvements as this is the most important benefit of using an LTS release, for instance, if you use Java 11 and the JDK version used in production is 11.0.1, then you are missing more than two years of important security, stability, and performance fixes defeating the purpose of LTS releases.

Reviewer: Could you add a blurb about what a Patch Update is prior to mentioning it here? Thanks!

> The main recomendation is that the JDK is updated with each Patch Update, when using an LTS release you should always test and update to the latest patch which contain security vulnerability fixes and other improvements as this is the most important benefit of using an LTS release, for instance, if you use Java 11 and the JDK version used in production is 11.0.1, then you are missing more than two years of important security, stability, and performance fixes defeating the purpose of LTS releases.

Reviewer: ```suggestion
The main recommendation is to keep JDK installs up to date with all patch updates to ensure the latest security vulnerability fixes and other improvements are available. For example, if you use Java 11 and the JDK version used in production is 11.0.1, then you are missing more than two years of important security, stability, and performance fixes defeating the purpose of LTS releases.
```
